src/Native/LibTorchSharp/THSNN.cpp

```diff
@@ -499,19 +499,34 @@ Tensor THSNN_poisson_loss(const Tensor input, const Tensor target, const bool lo
     )
 }
 
-Optimizer THSNN_Adam_ctor(const Tensor* parameters, const int length, const double learnig_rate)
+Optimizer THSNN_Adagrad_ctor(const Tensor* parameters, const int length, const double learning_rate, const double lr_decay, const double weight_decay)
+{
+    auto params = toTensors<at::Tensor>((torch::Tensor**)parameters, length);
+    auto options = torch::optim::AdagradOptions(learning_rate).lr_decay(lr_decay).weight_decay(weight_decay);
```

Comment:
What happens about the optionality of these?  

We don't have any consistent way of passing across optional values across the C API, just adhoc uses of NULL, min values and so on

Reply:
We should remove the defaults when crossing the interop boundary.

Reply:
I'll do that next.

Reply:
OK yes in this case that can be done.  

In some other cases I've so far made the optional explicit, e.g. see 
* `THSTensor_avg_pool2d_backward` `divisor_override` 
* `THSTensor_fft` 
* `THSTensor_upsample_nearest1d` - `outputSize` and `scaleFactors`

and so on. There are about 25 explicit uses of `c10::optional`

If you think it's better to simplify these out to optionality on the C# side please go ahead